common/optimiser/Optimiser.java

```diff
@@ -78,8 +87,12 @@ public boolean isError() {
         return status == Status.ERROR;
     }
 
+    private boolean constraintsChanged() {
+        return constraints.stream().anyMatch(c -> !c.isInitialised);
```

Comment:
something here is odd I will look into it after lunch in the next part of the review!